test/screenshot/screenshot.js

```diff
@@ -1,60 +1,209 @@
-      const imagePath = `./test/screenshot/${this.imagePath_}`;
-      await this.createScreenshotTask_(url, imagePath);
+      const golden = await this.takeScreenshot_();
+      const goldenHash = this.generateImageHash_(golden);
+      const goldenPath = this.getImagePath_(goldenHash, 'golden');
+      await Promise.all([
+        this.saveImage_(goldenPath, golden),
+        this.saveGoldenHash_(goldenHash),
+      ]);
       return;
     });
   }
 
+  /**
+   * Diffs a screenshot of the test URL with the existing golden screenshot
+   */
   diff() {
     test(this.urlPath_, async () => {
-      const url = `http://localhost:8080/${this.urlPath_}`;
-      const imagePath = `./test/screenshot/${this.imagePath_}`;
-      const snapshotImagePath = `./test/screenshot/${this.snapshotImagePath_}`;
-      const diffPath = `./test/screenshot/${this.diffPath_}`;
-
-      const [newScreenshot, oldScreenshot] = await Promise.all([
-        this.createScreenshotTask_(url, snapshotImagePath),
-        readFilePromise(imagePath),
+      // Get the golden file path from the golden hash
+      const goldenHash = await this.getGoldenHash_();
+      const goldenPath = this.getImagePath_(goldenHash, 'golden');
+
+      // Take a snapshot and download the golden iamge
+      const [snapshot, golden] = await Promise.all([
+        this.takeScreenshot_(),
+        this.readImage_(goldenPath),
       ]);
 
-      const data = await compareImages(newScreenshot, oldScreenshot,
-        comparisonOptions);
+      // Compare the images
+      const data = await compareImages(snapshot, golden, comparisonOptions);
+      const diff = data.getBuffer();
+
+      // Use the same hash for the snapshot path and diff path so it's easy can associate the two
+      const snapshotHash = this.generateImageHash_(snapshot);
+      const snapshotPath = this.getImagePath_(snapshotHash, 'snapshot');
+      const diffPath = this.getImagePath_(snapshotHash, 'diff');
+      const metadata = {golden: goldenHash};
+
+      // Save the snapshot and the diff
+      await Promise.all([
+        this.saveImage_(snapshotPath, snapshot, metadata),
+        this.saveImage_(diffPath, diff, metadata),
+      ]);
+
+      return assert.isBelow(Number(data.misMatchPercentage), 0.01);
+    });
+  }
+
+  /**
+   * Generates a unique hash from an image's contents
+   * @param {!Buffer} imageBuffer The image buffer to hash
+   * @return {string}
+   * @private
+   */
+  generateImageHash_(imageBuffer) {
+    return createHash('sha256').update(imageBuffer).digest('hex');
+  }
+
+  /**
+   * Returns the golden hash
+   * @return {string|undefined}
+   * @private
+   */
+  async getGoldenHash_() {
+    const goldenFile = await readFilePromise(goldenFilePath);
+    const goldenJSON = JSON.parse(goldenFile);
+    return goldenJSON[this.urlPath_];
+  }
+
+  /**
+   * Returns the correct image path
+   * @param {string} imageHash The image hash
+   * @param {string} imageType The image type
```

Comment:
Can we make an `@enum` for this? It can have string values so that your existing usages will continue to work.